README
Securechat is a text-based client/server chat utility  that uses ssl/tls (openssl)  to establish an encrypted communication between the two parties. It wants to be a "modern" version of the talk program, the historical text chat program on Unix.
The program requires the generation of some certificates to operate. Consult the website (www.dynamiclib.com) and the man page for detailed instructions.



ChangeLog
securechat (0.8.41) any; urgency=low

  * Beta Test
    Added makefile and spec file to generate AIX rpm. Some minor cleaning.
 -- Gabriele Bonacini <gabriele.bonacini@protonmail.com>  Wed, 15 Jan 2014 02:59:00 -0100

securechat (0.8.4) any; urgency=low

  * Beta Test
    Cleaning.
 -- Gabriele Bonacini <gabriele.bonacini@protonmail.com>  Sat, 11 Jan 2014 19:41:00 -0100

securechat (0.8.3) any; urgency=low

  * Beta Test 
    First pubblic release.
 -- Gabriele Bonacini <gabriele.bonacini@protonmail.com>  Sat, 11 Jan 2014 19:41:00 -0100

securechat (0.8.2) any; urgency=low

  * Beta Test 
    Added cofiguration scripts.
 -- Gabriele Bonacini <gabriele.bonacini@protonmail.com>  Sat, 11 Jan 2014 02:26:00 -0100

securechat (0.8.0) any; urgency=low

  * Beta Test 
    Added minimal desired functions.
 -- Gabriele Bonacini <gabriele.bonacini@protonmail.com>  Fri, 13 Sep 2013 12:24:00 -0100

securechat (0.7.5) any; urgency=low

  * Beta Test 
    Removed major bugs.
 -- Gabriele Bonacini <gabriele.bonacini@protonmail.com>  Mon, 26 Aug 2013 22:34:00 -0100

securechat (0.6.2) any; urgency=low

  * Beta Test 
    Added all basic functions.
 -- Gabriele Bonacini <gabriele.bonacini@protonmail.com>  Mon, 19 Aug 2013 13:03:00 -0100

securechat (0.5.0) any; urgency=low

  * Beta Test 
    The beta version is ready.
 -- Gabriele Bonacini <gabriele.bonacini@protonmail.com>  Thu, 25 Jul 2013 05:26:00 -0100
This file is present for compatibility with some scripts. For the actual changes, see "changelog".


src/chat.c
// -----------------------------------------------------------------
// Securechat - A simple chat program over SSL/TLS.
// Copyright (C) 2013  Gabriele Bonacini
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software Foundation,
// Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
// -----------------------------------------------------------------

#include <chat.h>

volatile sig_atomic_t toRefresh;
bool debugOn=false;
int logDescriptor;
FILE *logStream;

int main(int argc, char **argv){
	struct chatContext context=CONTEXT_ZERO;
	struct termios backupTerm;

	bool quit=false;
	int current=0;
	char version[]=VERSION; 

	pthread_t readerId=0;
	pthread_t listenerId=0;
	pthread_t refresherId=0;
	pthread_attr_t readerAttr;
	pthread_attr_t listenerAttr;
	pthread_attr_t refresherAttr;

	if(!checkParameters(argc,argv)){
		exit(EXIT_FAILURE);
	}

	if (tcgetattr(STDIN_FILENO, &backupTerm) == -1){
		perror("Can't save terminal settings.");
		exit(EXIT_FAILURE);
	}

	if(initscr() == NULL){
		perror("Init screen");
		exit(EXIT_FAILURE);
	} 	

	if(LINES<MINLINES || COLS<MINCOLUMNS){
		perror(WRONG_SIZE);
		(void)endwin();			
		exit(EXIT_FAILURE);
	}

	setDebug();

	(void)signal(SIGWINCH, resizeHandler);

	initContext(&context);

	setTerm();

	SSL_load_error_strings();
	ERR_load_BIO_strings();
	ERR_load_SSL_strings();
	OpenSSL_add_all_algorithms();
	(void)SSL_library_init();

	printMenu(UPPER_MENU);
	changeStatus(context.statusDisconnectp, context.statusDisconnectSize);
	printVersion(version,strlen(version));

	setWindows(&context);

	(void)(void)pthread_attr_init(&readerAttr);
	(void)pthread_attr_setdetachstate(&readerAttr, PTHREAD_CREATE_DETACHED);
	(void)pthread_create(&readerId, &readerAttr, &readIncoming, &context);

	(void)pthread_attr_init(&listenerAttr);
	(void)pthread_attr_setdetachstate(&listenerAttr, PTHREAD_CREATE_DETACHED);

	(void)pthread_attr_init(&refresherAttr);
	(void)pthread_attr_setdetachstate(&refresherAttr, PTHREAD_CREATE_DETACHED);
	(void)pthread_create(&refresherId, &refresherAttr, &refreshAll, &context);

	while(!quit){
		current=wgetch(context.down_winp);
		switch(current){
			case KEY_F(5):
				quit=true;
			break;
			case KEY_F(6):
				manageConfigurationModule(&context, &listenerId, &listenerAttr);
			break;
			case C_NEWLINE: 
				sendMessage(&context);
			break;
			case C_BACKSPACE:
			case KEY_BACKSPACE:
				doBackspace(&context);
			break;
			case KEY_LEFT:
				moveBackward(&context);
			break;
			case KEY_RIGHT:
				moveForward(&context);
			break;
			case KEY_UP:
				moveBackward(&context);
			break;
			case KEY_DOWN:
				moveForward(&context);
			break;
			default: 
				parseInput(&context, current);

		}
	}

	(void)pthread_cancel(readerId);
	(void)pthread_cancel(refresherId);
	(void)pthread_cancel(listenerId);

	(void)clear();	
	printMenu("BYE!");	
	(void)sleep(1);

	freeFormContext(&context);

	(void)endwin();			

	disconnectNotify(&context);

	cleanContext(&context);

	(void)tcsetattr(STDIN_FILENO, TCSANOW, &backupTerm);

	endDebug();

	return 0;
}


include/chat.h
#ifndef _CHAT_H_
#define _CHAT_H_

// -----------------------------------------------------------------
// Securechat - A simple chat program over SSL/TLS.
// Copyright (C) 2013  Gabriele Bonacini
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software Foundation,
// Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
// -----------------------------------------------------------------

#include <config.h>

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>
#include <fcntl.h>
#include <form.h>
#include <openssl/bio.h>
#include <openssl/ssl.h>
#include <openssl/err.h>
#include <pthread.h>
#include <signal.h>
#include <sys/ioctl.h>
#include <term.h>
#include <termios.h>
#include <stdarg.h>

//AIX

#ifdef HAVE_STDBOOL_H 
#include <stdbool.h>
#endif

#define C_BACKSPACE 127
#define C_NEWLINE '\n'
#define C_TAB '\t'

#define UP_HEIGHT ( abs(LINES-(LINES / 4 )) - 2)
#define UP_WIDTH (COLS - 2)
#define DOWN_HEIGHT ( abs((LINES / 4 )) - 3 )
#define DOWN_WIDTH (COLS - 2)
#define DOWN_Y (abs(LINES-(LINES / 4 )) + 1)
#define DOWN_X 1
#define UP_Y 1
#define UP_X 1

#define MINLINES 35
#define MINCOLUMNS 95
#define WRONG_SIZE "Screen too small, minimum dimensions are 35 lines and 95 columns."

#define PADDING 1
#define STATUS_OFFSET 5
#define VERSION_OFFSET 15

#define INSERT_FORM_FIELDS 2
#define HISTORY_FORM_FIELDS 2
#define CONFIGURATION_FORM_FIELDS 24
#define SMALL_BUFFER 64
#define MEDIUM_BUFFER 256
#define BIG_BUFFER 2048
#define HUGE_BUFFER 10240

#define STATUS_CONNECTED "STATUS: CONNECTED          "
#define STATUS_DISCONNECTED "STATUS: DISCONNECT/INACTIVE"
#define STATUS_LISTENING "STATUS: LISTENING          "

#define UPPER_MENU "  - F5 Exit - F6 Configure - <Arrows> Scroll the log "

#define ERROR_PROMPT "Error:---> "
#define INFO_PROMPT "Info:---> "
#define LOCAL_PROMPT "Local:---> "
#define REMOTE_PROMPT "Remote:---> "

#define LOGFILE "/tmp/securechat.log"

#define POLLING_INTERVAL 100000

#define BLACKLIST "SCBLACKLIST"
#define DEFAULT_BLACKLIST "!aNULL:!eNULL:!3DES:!SHA1:!EXPORT:!EXPORT56:MEDIUM"

enum status { inactive, connected, listening };

struct chatContext{
	bool serverMode;
	int wrows;
	int wcols;
	int up_height;
	int up_width;
	int localPort;
	int remotePort;
	char* baseDir;
	BIO *biop;
	BIO *abiop;
	BIO *mbiop;
	SSL_CTX *ctxp;
	SSL *sslp;
	WINDOW *up_winp;
	WINDOW *down_winp;
	WINDOW *inner_up_winp;
	WINDOW *inner_down_winp;
	WINDOW *configurationp;
	WINDOW *inner_configurationp;
	FORM  *historyFormp;
	FORM  *configurationFormp;
	FORM  *insertFormp;
	FIELD *historyFieldp[HISTORY_FORM_FIELDS];
	FIELD *insertFieldp[INSERT_FORM_FIELDS];
	FIELD *configurationFieldp[CONFIGURATION_FORM_FIELDS];
	char incomingBufferp[MEDIUM_BUFFER];
	size_t incomingBufferSize;
	const char* statusConnectp;
	size_t statusConnectSize;
	const char* statusDisconnectp;
	size_t statusDisconnectSize;
	const char* statusListeningp;
	size_t statusListeningSize;
	char configIP[SMALL_BUFFER];
	size_t configIPSize;
	char sConfigPort[SMALL_BUFFER];
	size_t sConfigPortSize;
	char errBuffer[MEDIUM_BUFFER];
	size_t errBufferSize;
	char password[MEDIUM_BUFFER];
	size_t passwordSize;
	char handShakeSummary[MEDIUM_BUFFER];
	size_t handShakeSummarySize;
        int labelIpAddress;
        int labelPortAddress;
	int startIPAddress;
	int valuePortAddress;
	int passworHeaderAddress;
	int passwordAddress;
	int passworFooterAddress;
	int passworConfirmHeaderAddress;
	int passwordConfirmAddress;
	int passworConfirmFooterAddress;
	int serverModeAddress;
	int passwdErrorAddress;
	/* Status: Valid values: inactive, connected, listening. */
	enum status status;
	pthread_mutex_t historyFieldMtxp;
	char *blackList;
};

#define CONTEXT_ZERO {  false, /*bool serverMode */\
			0, /* int wrows */\
			0, /* int wcols */\
			0, /* int up_height */\
			0, /* int up_width */\
			0, /* int localPort */\
			0, /* int remotePort */\
			NULL, /* char* baseDir */\
			NULL, /* BIO *biop */\
			NULL, /* BIO *abiop */\
			NULL, /* BIO *mbiop */\
			NULL, /* SSL_CTX *ctxp */\
			NULL, /* SSL *sslp */\
			NULL, /* WINDOW *up_winp */\
			NULL, /* WINDOW *down_winp */\
			NULL, /* WINDOW *inner_up_winp */\
			NULL, /* WINDOW *inner_down_winp */\
			NULL, /* WINDOW *configurationp */\
			NULL, /* WINDOW *inner_configurationp */\
			NULL, /* FORM  *historyFormp */\
			NULL, /* FORM  *configurationFormp */\
			NULL, /* FORM  *insertFormp */\
			{0}, /* FIELD *historyFieldp[HISTORY_FORM_FIELDS] */\
			{0}, /* FIELD *insertFieldp[INSERT_FORM_FIELDS] */\
			{0}, /* FIELD *configurationFieldp[CONFIGURATION_FORM_FIELDS] */\
			{0}, /* char incomingBufferp[MEDIUM_BUFFER] */\
			(size_t)0, /* size_t incomingBufferSize */\
			STATUS_CONNECTED, /* const char* statusConnectp */\
			(size_t)0, /* size_t statusConnectSize */\
			STATUS_DISCONNECTED, /* const char* statusDisconnectp */\
			(size_t)0, /* size_t statusDisconnectSize */\
			STATUS_LISTENING, /* const char* statusListeningp */\
			(size_t)0, /* size_t statusListeningSize */\
			{0}, /* char configIP[SMALL_BUFFER] */\
			(size_t)0, /* size_t configIPSize */\
			{0}, /* char sConfigPort[SMALL_BUFFER] */\
			(size_t)0, /* size_t sConfigPortSize */\
			{0}, /* char errBuffer[MEDIUM_BUFFER] */\
			(size_t)0, /* size_t errBufferSize */\
			{0}, /* char password[MEDIUM_BUFFER] */\
			(size_t)0, /* size_t passwordSize */\
			{0}, /*char handShakeSummary[MEDIUM_BUFFER]*/\
			(size_t)0, /*size_t handShakeSummarySize*/\
			0, /* int labelIpAddress */\
			0, /* int labelPortAddress */\
			0, /* int startIPAddress */\
			0, /* int valuePortAddress */\
			0, /* int passworHeaderAddress */\
			0, /* int passwordAddress */\
			0, /* int passworFooterAddress */\
			0, /* int passworConfirmHeaderAddress */\
			0, /* int passwordConfirmAddress */\
			0, /* int passworConfirmFooterAddress */\
			0, /* int serverModeAddress */\
			0, /* int passwdErrorAddress */\
			inactive, /* enum status status */\
			PTHREAD_MUTEX_INITIALIZER, /* pthread_mutex_t historyFieldMtxp */\
			NULL /* char *blackList;*/\
}

WINDOW *new_win(int height, int width, int starty, int startx);
void free_win(WINDOW *local_win);
void setTerm(void);

bool checkParameters(int argc, char **argv);

void changeStatus(const char* string, size_t size);
void printMenu(const char* string);
void printVersion(const char* string, size_t size);
void printHistory(const char* prompt, const char* errorMessage, struct chatContext *context);

// Threads 
/*@null@*/
void* readIncoming(void* chatContext);
/*@null@*/
void* listenIncoming(void* chatContext);
int password_callback(char *buf, int size, int rwflag, void *chatContext);
/*@null@*/
void* refreshAll(void* chatContext);
void disconnectNotify(struct chatContext *context);

void sendMessage(struct chatContext *context);
void parseInput(struct chatContext *context, int current);
void doBackspace(struct chatContext *context);

void resizeHandler(int sig);

void initContext(struct chatContext *context);
bool setContext(struct chatContext *context);
void cleanContext(struct chatContext *context);
void freeFormContext(struct chatContext *context);

void setWindows(struct chatContext *context);
void showConfigWin(struct chatContext *context);
void setConfigurationForm(struct chatContext *context);
void manageConfigurationModule(struct chatContext *context, pthread_t *listenerId, pthread_attr_t *listenerAttr);
void extractConfiguration(struct chatContext *context, char *configReport);
bool passwordsMatch(struct chatContext *context);
void notifyPasswordError(struct chatContext *context);
void setServerMenu(bool activation, struct chatContext *context);

bool setClientMode(struct chatContext *context);
bool setServerMode(struct chatContext *context);
void moveForward(struct chatContext *context);
void moveBackward(struct chatContext *context);
void trimTrailingSpaces(char *buffer, size_t size);

void setDebug(void);
void writeSslError(const char *errString);
void writeErrorLog(const char *errString, ...);
void endDebug(void);

#endif
